### PR TITLE
Fixes bugs in pyspark @with_columns

### PR DESCRIPTION
**Pull request by skrawcz**
1. Configuration linkage was not exposed. So adding a parameter for people to specify what configuration is required.
2. There was a bug in the UDF logic, because `all(empty)` evaluates to True. So fixed that.
3. The `select` in append mode appended all possible columns. It didn't only append what was in the `select`. This changes that behavior by dropping "outputs" from the subdag that aren't in the select. We do it this way because we don't know what's in the original dataframe.

## Changes
- h_spark
- llm workflow example using the PDF summarizer

## How I tested this
 - runs locally

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

----

| <a href="https://ellipsis.dev" target="_blank">Ellipsis</a> | :rocket: This PR description was created by [Ellipsis](https://www.ellipsis.dev) for commit 9341d3147e2c7956f8136f7019d21aafa7983aec.  | 
|--------|--------|

### Summary:
This PR enhances the pyspark @with_columns decorator, fixes a bug in the UDF logic, modifies the 'select' behavior in append mode, updates the example workflows, and adds new files in the `pdf_summarizer/run_on_spark` directory.

**Key points**:
- Enhanced the pyspark @with_columns decorator by adding a parameter for required configurations.
- Fixed a bug in the UDF logic where `all(empty)` was incorrectly evaluating to True.
- Modified the 'select' behavior in append mode to only append the specified columns.
- Updated the example workflows to demonstrate the new features and fixes.
- Added new files `run_with_columns.py` and `spark_pdf_pipeline.py` in the `pdf_summarizer/run_on_spark` directory.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: examples/LLM_Workflows/pdf_summarizer/run_on_spark/spark_pdf_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import pandas as pd
+import pyspark.sql as ps
+import summarization
+
+from hamilton.plugins.h_spark import with_columns
+
+
+def pdf_df(spark_session: ps.SparkSession) -> ps.DataFrame:
+    pandas_df = pd.DataFrame(
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply examples/LLM_Workflows/pdf_summarizer/run_on_spark/spark_pdf_pipeline.py:9]: this is an example related to the file in the example folder.

File: hamilton/plugins/h_spark.py [hunk 2]
```diff
@@ -1087,6 +1089,29 @@ def new_callable(**kwargs) -> DataFrame:
             input_types={upstream_name: DataFrame},
         )
 
+    @staticmethod
+    def create_drop_node(
```
Comment (elijahbenizzy) [inline hamilton/plugins/h_spark.py:1093]: Other possibility is for this to depend on the input dataframe as well and ask it for its columns, then add them to select. This seems clearer though.

**Review by elijahbenizzy — COMMENTED** [review]
Looks good -- made a few suggestions.

File: hamilton/plugins/h_spark.py [hunk 3]
```diff
@@ -1014,11 +1015,12 @@ def final_df(initial_df: ps.DataFrame) -> ps.DataFrame:
...
         :param mode: The mode of the operation. This can be either "append" or "select".
-            If it is "append", it will keep all columns in the dataframe. If it is "select",
+            If it is "append", it will keep all original columns in the dataframe. If it is "select",
             it will only keep the columns in the dataframe from the `select` parameter. Note that,
-            if the `select` parameter is left blank, it will keep all columns in the dataframe
-            that are in the subdag (as that is the behavior of the `select` parameter. This
-            defaults to `append`
+            if the `select` parameter is left blank, it will add all columns in the dataframe
+            that are in the subdag. This defaults to `append`.
+        :param config_required: the list of config keys that are required to resolve any functions. Pass in None\
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/plugins/h_spark.py:1022]: yes/no.  Lets go with more constrained, we can always relax it.

File: plugin_tests/h_spark/test_h_spark.py [hunk 4]
```diff
@@ -607,6 +608,27 @@ def df_as_pandas(df: DataFrame) -> pd.DataFrame:
     assert set(nodes_by_names.keys()) == {"df_as_pandas.c", "df_as_pandas"}
 
 
+def test_with_columns_generate_nodes_select_append_mode():
+    dec = h_spark.with_columns(
+        a_dag,
+        columns_to_pass=["input"],
+        select=["c"],
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply plugin_tests/h_spark/test_h_spark.py:638]: added select

**Comment by skrawcz** [comment]
TODO:
 - redo commits
 - make sure comments/docs what we want to show people

**Review by elijahbenizzy — APPROVED** [review]
Looks good

**Pull request merged**